File: 516_LongestPalindromicSubsequence.py
class Solution:
    def longestPalindromeSubseq(self, s: str) -> int:
        ## 2d dp
        length = len(s)
        
        table = [[0] * length for _ in range(length)]
        for i in range(length):
            table[i][i] = 1
            
        for i in range(length - 2, -1, -1):
            for j in range(i + 1, length):
                # print (f'{i}, {j}')
                if s[i] == s[j]:
                    table[i][j] = max(table[i + 1][j], table[i][j - 1], table[i + 1][j - 1] + 2)
                else:
                    table[i][j] = max(table[i + 1][j], table[i][j - 1])
        return table[0][length - 1]
        
        ## 1d dp
        table = [1] * length
        for i in range(length - 2, -1, -1):
            diag_val = 0
            for j in range(i + 1, length):
                # print (f'{i}, {j}')
                if s[i] == s[j]:
                    new_val = max(table[j], table[j - 1], diag_val + 2)
                else:
                    new_val = max(table[j], table[j - 1])
                diag_val = table[j]
                table[j] = new_val
                
        return table[length - 1]
        

File: 516_LongestPalindromicSubsequence.cpp
class Solution {
public:
    int longestPalindromeSubseq(string s) {

        if (s == "") return 0;
        size_t n = s.size();
        int f[n][n];
        fill_n(&f[0][0], n*n, 0);

        // vector<vector<int> > f(n, vector<bool>(n, false));
        int start = 0, max_length = 1;

        for (int i = 0; i < n; ++i) f[i][i] = 1;

        for (int i = n - 2; i > -1; --i) {
            for (int j = i + 1; j < n; ++j) {
                if (s[i] == s[j]) f[i][j] = f[i + 1][j - 1] + 2;
                else f[i][j] = max(f[i + 1][j], f[i][j - 1]);

            }
        }
        return f[0][n - 1];
    }
};